### ISSUE TITLE
Change inferred return type of `async def` from `typing.Coroutine` to `types.CoroutineType`

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#9849]
An `async def` function returns a type of `types.CoroutineType` at runtime. This is a subtype of `typing.Coroutine`, which is the type currently inferred. It would be better to use the more accurate type here.

This was pointed out in [this mypy issue](https://github.com/python/mypy/issues/18635).

**Comment by rafaelclp** [comment]
@erictraut This broke a lot of stuff on our side. We have decorators that expect an async function and return an async function. Now they break when used in methdos that override methods from an ABC:

```python
def decorator(fn: Callable[P, Coroutine[Any, Any, RT]]) -> Callable[P, Coroutine[Any, Any, RT]]: ...

class Base(ABC):
    @abstractmethod
    async def do_something(self) -> X: ...

class Derived(Base):
    @decorator
    async def do_something(self) -> X: ...
```

We get this error:
```
    Return type mismatch: base method returns type "CoroutineType[Any, Any, X]", override returns type "Coroutine[Any, Any, X]"
      "Coroutine[Any, Any, X]" is not assignable to "CoroutineType[Any, Any, X]" (reportIncompatibleMethodOverride)
```

Should we change all our decorators to take `Callable[P, CoroutineType[Any, Any, RT]]` instead?

**Comment by erictraut** [comment]
> Should we change all our decorators to take `Callable[P, CoroutineType[Any, Any, RT]]` instead?

That's one way to make your program type check without errors. Another (perhaps preferable) way is to change your abstract base class to indicate that `do_something` returns a `Coroutine[Any, Any, X]` rather than an (implied) `CoroutineType[Any, Any, X]`.

```python
class Base(ABC):
    @abstractmethod
    def do_something(self) -> Coroutine[Any, Any, X]: ...
```

**Comment by rafaelclp** [comment]
I changed the base class, but then Pylint (in its latest version, 3.3.6) produces its own warning:
```
W0236: Method 'do_something' was expected to be 'non-async', found it instead as 'async' (invalid-overridden-method)
```

I just disabled `invalid-overridden-method` in my `.pylintrc`. We are slowly migrating all repos to use ruff instead anyway...

I think this change to mypy and pyright was a bit premature though.

- If you search how to typehint async functions in python, you won't find anyone or any docs telling you to use `CoroutineType`, only `Coroutine` or `Awaitable`.
- This means most libs that typehint async decorators use `Coroutine` or `Awaitable`. They will not work on methods that override async methods. People will get confused, and won't know how to fix it.
- Even if people do find out how to fix it, many still use Pylint, which will complain with the error above. Of course, this should be fixed in Pylint, not in Pyright or Mypy; but in the meantime, users will be affected nonetheless.
- Even after it's fixed in Pylint, and after people learn how to get around the issue with Pyright, they won't be able to confidently use `async def f(...) ->X` abstract methods (which is much more readable than `def f(...) -> Coroutine[Any, Any, X]`), as that might not be compatible with decorators in some libs down the line.

It might be better, in theory, to use the most accurate type in pyright and mypy. But from a practical stadnpoint, I'm not so sure that's the right choice.

**Comment by tuure-vincit** [comment]
I am also confused by this change. I tried changing the type hints in my project from `Coroutine[Any, Any, RT]` to `CoroutineType[Any, Any, RT]` but now I get runtime errors when running tests complaining that `TypeError: type 'coroutine' is not subscriptable`. Did I do something wrong? How am I supposed to type hint these decorator functions in a way that avoids both type checker errors and runtime errors?

Furthermore, @erictraut, you had earlier expressed doubts about changing these type checks [here](https://github.com/python/typing/issues/1480). Relevant quote below:

> The classes defined in types are runtime details. Unless you're writing tooling-related code that needs access to these details (e.g. a runtime type checker or debugger), you should not use these classes in type annotations. This is already a point of confusion for newcomers to typing, and my blanket guidance is "if you're importing a type from the types module and using it in a type annotation, you're probably doing something wrong." If we start to promote the use of some of the classes from the types module, it will create significant confusion. It will also require type checkers to expand existing special-case logic that knows about classes in the typing and collections.abc modules. 

So, why the change of mind? I feel that this change has definitely made using `Coroutine`/`CoroutineType` in type hints more confusing...

**Comment by tuure-vincit** [comment]
Here's a minimal code example of the issue I'm running into:

```python
import asyncio
import functools
from abc import ABC, abstractmethod
from types import CoroutineType  # pyright: ignore[reportUnusedImport]  # noqa: F401
from typing import (
    Any,
    Callable,
    Coroutine,  # pyright: ignore[reportUnusedImport]  # noqa: F401
    ParamSpec,
    TypeVar,
    override,
)

R = TypeVar("R")
P = ParamSpec("P")


def decorator(
    func: Callable[P, CoroutineType[Any, Any, R]],
) -> Callable[P, CoroutineType[Any, Any, R]]:
    @functools.wraps(func)
    def wrapper(*args: P.args, **kwargs: P.kwargs) -> CoroutineType[Any, Any, R]:
        return func(*args, **kwargs)

    return wrapper


class TestABC(ABC):
    @abstractmethod
    async def test_func(self, data: int) -> int:
        pass


class TestClass(TestABC):
    @decorator
    @override
    async def test_func(self, data: int) -> int:
        return data


async def main():
    test_object = TestClass()
    result = await test_object.test_func(1)
    print(result)


if __name__ == "__main__":
    loop = asyncio.new_event_loop()
    asyncio.set_event_loop(loop)
    asyncio.run(main())

```

The way it's currently set up, I do not get any error's from pyright, but if I run this code, I get `TypeError: type 'coroutine' is not subscriptable` from line 19. Alternatively, I can change all mentions of `CoroutineType` to `Coroutine` instead, at which points the code runs, but then I get a pyright type checking error `Method "test_func" overrides class "TestABC" in an incompatible manner` from line 37...
